### PR TITLE
Implement support for time-to-live for map keys

### PR DESCRIPTION
**Pull request by kuujo**
This PR implements support for time-to-live durations for map keys, e.g.

```java
ConsistentMap<String, String> map = atomix.consistentMapBuilder("foo").build();
map.put("foo", "bar", Duration.ofSeconds(1));
Thread.sleep(2);
assert map.get("foo") == null;
```

### CODE REVIEW

File: core/src/main/java/io/atomix/core/map/AsyncConsistentMap.java [hunk 1]
```diff
@@ -213,7 +212,35 @@ default PrimitiveType primitiveType() {
    * @return the previous value (and version) associated with key, or null if there was
    * no mapping for key.
    */
-  CompletableFuture<Versioned<V>> put(K key, V value);
+  default CompletableFuture<Versioned<V>> put(K key, V value) {
+    return put(key, value, Duration.ZERO);
+  }
+
+  /**
+   * Associates the specified value with the specified key in this map (optional operation).
```
Comment (johnou) [inline core/src/main/java/io/atomix/core/map/AsyncConsistentMap.java]: Is optional operation referring to the TTL?

File: core/src/main/java/io/atomix/core/map/impl/ConsistentMapService.java [hunk 2]
```diff
@@ -342,6 +344,43 @@ protected boolean valueIsNull(MapEntryValue value) {
...
+
+  /**
+   * Schedules the TTL for the given value.
+   *
+   * @param value the value for which to schedule the TTL
+   */
+  protected void scheduleTtl(String key, MapEntryValue value) {
+    if (value.ttl() > 0) {
+      value.timer = scheduler().schedule(Duration.ofMillis(value.ttl() - value.created()), () -> {
+        entries().remove(key);
```
Comment (johnou) [inline core/src/main/java/io/atomix/core/map/impl/ConsistentMapService.java]: Wouldn't remove(key, value) be safer?

Comment (kuujo) [reply core/src/main/java/io/atomix/core/map/impl/ConsistentMapService.java]: Yeah good idea

File: core/src/main/java/io/atomix/core/map/AsyncConsistentMap.java (hunk 1 above)
Comment (kuujo) [reply core/src/main/java/io/atomix/core/map/AsyncConsistentMap.java]: `Duration.ZERO` creates no TTL

**Pull request merged**